File: src/rust/laboneq-dsl/src/signal_calibration/port_mode.rs
// Copyright 2026 Zurich Instruments AG
// SPDX-License-Identifier: Apache-2.0

use std::str::FromStr;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum PortMode {
    LF,
    RF,
}

impl FromStr for PortMode {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "lf" => Ok(PortMode::LF),
            "rf" => Ok(PortMode::RF),
            _ => Err(format!("Unknown port mode: {}", s)),
        }
    }
}

impl std::fmt::Display for PortMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            PortMode::LF => "LF",
            PortMode::RF => "RF",
        };
        write!(f, "{}", s)
    }
}

File: src/python/laboneq/core/types/enums/port_mode.py
# Copyright 2022 Zurich Instruments AG
# SPDX-License-Identifier: Apache-2.0

from enum import Enum


class PortMode(Enum):
    LF = "LF"
    RF = "RF"

    @classmethod
    def _missing_(cls, value):
        if isinstance(value, str) and value != value.upper():
            return cls(value.upper())
        return None
